File: generateExplanations.js
const { createClient } = require('@supabase/supabase-js');
const Anthropic = require('@anthropic-ai/sdk');

const supabaseUrl = 'https://rabavobdklnwvwsldbix.supabase.co';
const supabaseKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InJhYmF2b2Jka2xud3Z3c2xkYml4Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTk3NjU0NzgsImV4cCI6MjA3NTM0MTQ3OH0.z_1N3TG-cS9Bc1s7UAif91PkIjhBKvicrUqupiNP80Y';

const supabase = createClient(supabaseUrl, supabaseKey);

// This will be populated from your environment or passed as argument
const ANTHROPIC_API_KEY = process.env.ANTHROPIC_API_KEY;

async function generateExplanation(question, passage, questionType) {
  const client = new Anthropic({ apiKey: ANTHROPIC_API_KEY });

  const prompt = `You are writing a SIMPLE, CLEAR explanation for an ACT ${questionType} test question.

PASSAGE:
${passage}

QUESTION: ${question.question_text}

CHOICES:
${typeof question.choices === 'string' ? JSON.parse(question.choices).join('\n') : question.choices.join('\n')}

CORRECT ANSWER: ${question.correct_answer}

Write an explanation in this EXACT HTML format:

<div style="line-height: 1.6; margin-bottom: 0.75rem; color: #374151;">
[1-2 sentences explaining why the correct answer is right. For Reading: reference what the passage says. For Science: cite specific data (Table X, Figure Y, values). Keep it brief and clear.]
</div>

<div>
<strong style="font-size: 0.875rem; color: #6b7280;">Why Other Answers Are Wrong:</strong>
<div style="margin-top: 0.5rem; line-height: 1.6; color: #374151;">
${getWrongChoiceTemplate(question.correct_answer, question.choices)}
</div>
</div>

Requirements:
- Total length: 600-900 characters
- Use simple language
- Be specific and reference the ${questionType === 'Reading' ? 'passage' : 'data'}
- Keep it concise`;

  const message = await client.messages.create({
    model: 'claude-3-5-sonnet-20241022',
    max_tokens: 1024,
    messages: [{
      role: 'user',
      content: prompt
    }]
  });

  return message.content[0].text;
}

function getWrongChoiceTemplate(correctAnswer, choices) {
  const choicesList = typeof choices === 'string' ? JSON.parse(choices) : choices;
  const letters = ['A', 'B', 'C', 'D', 'F', 'G', 'H', 'J'];

  return choicesList
    .map((choice, idx) => {
      const letter = choice.match(/^([A-Z]):/)?.[1] || letters[idx];
      if (letter === correctAnswer) return null;
      return `<div style="margin-bottom: 0.375rem;"><strong>Choice ${letter}:</strong> [Why wrong]</div>`;
    })
    .filter(Boolean)
    .join('\n');
}

async function processReadingQuestions() {
  console.log('Fetching reading questions and passages...');

  const { data: questions, error: qError } = await supabase
    .from('practice_test_reading_questions')
    .select('*')
    .order('id', { ascending: true })
    .limit(40);

  if (qError) {
    console.error('Error fetching questions:', qError);
    return;
  }

  // Get unique passage IDs
  const passageIds = [...new Set(questions.map(q => q.passage_id))];

  const { data: passages, error: pError } = await supabase
    .from('practice_test_reading_passages')
    .select('*')
    .in('id', passageIds);

  if (pError) {
    console.error('Error fetching passages:', pError);
    return;
  }

  // Create passage lookup
  const passageMap = {};
  passages.forEach(p => {
    passageMap[p.id] = p.passage_text;
  });

  console.log(`Processing ${questions.length} reading questions...`);

  for (let i = 0; i < questions.length; i++) {
    const question = questions[i];
    const passage = passageMap[question.passage_id];

    console.log(`\nProcessing question ${i + 1}/${questions.length} (ID: ${question.id})`);

    try {
      const explanation = await generateExplanation(question, passage, 'Reading');

      const { error: updateError } = await supabase
        .from('practice_test_reading_questions')
        .update({ explanation })
        .eq('id', question.id);

      if (updateError) {
        console.error(`Error updating question ${question.id}:`, updateError);
      } else {
        console.log(`✓ Updated question ${i + 1}`);
      }

      // Report progress every 15 questions
      if ((i + 1) % 15 === 0) {
        console.log(`\n========== PROGRESS: ${i + 1}/${questions.length} Reading questions complete ==========\n`);
      }

      // Small delay to avoid rate limits
      await new Promise(resolve => setTimeout(resolve, 1000));
    } catch (error) {
      console.error(`Error processing question ${question.id}:`, error);
    }
  }

  console.log('\n========== ALL READING QUESTIONS COMPLETE ==========\n');
}

async function processScienceQuestions() {
  console.log('Fetching science questions and passages...');

  const { data: questions, error: qError } = await supabase
    .from('practice_test_science_questions')
    .select('*')
    .order('id', { ascending: true })
    .limit(40);

  if (qError) {
    console.error('Error fetching questions:', qError);
    return;
  }

  // Get unique passage IDs
  const passageIds = [...new Set(questions.map(q => q.passage_id))];

  const { data: passages, error: pError } = await supabase
    .from('practice_test_science_passages')
    .select('*')
    .in('id', passageIds);

  if (pError) {
    console.error('Error fetching passages:', pError);
    return;
  }

  // Create passage lookup
  const passageMap = {};
  passages.forEach(p => {
    passageMap[p.id] = p.passage_text;
  });

  console.log(`Processing ${questions.length} science questions...`);

  for (let i = 0; i < questions.length; i++) {
    const question = questions[i];
    const passage = passageMap[question.passage_id];

    console.log(`\nProcessing question ${i + 1}/${questions.length} (ID: ${question.id})`);

    try {
      const explanation = await generateExplanation(question, passage, 'Science');

      const { error: updateError } = await supabase
        .from('practice_test_science_questions')
        .update({ explanation })
        .eq('id', question.id);

      if (updateError) {
        console.error(`Error updating question ${question.id}:`, updateError);
      } else {
        console.log(`✓ Updated question ${i + 1}`);
      }

      // Report progress every 15 questions
      if ((i + 1) % 15 === 0) {
        console.log(`\n========== PROGRESS: ${i + 1}/${questions.length} Science questions complete ==========\n`);
      }

      // Small delay to avoid rate limits
      await new Promise(resolve => setTimeout(resolve, 1000));
    } catch (error) {
      console.error(`Error processing question ${question.id}:`, error);
    }
  }

  console.log('\n========== ALL SCIENCE QUESTIONS COMPLETE ==========\n');
}

async function main() {
  console.log('Starting explanation generation...\n');

  // Process Reading questions first
  await processReadingQuestions();

  // Then Science questions
  await processScienceQuestions();

  console.log('\n========== ALL 80 EXPLANATIONS COMPLETE ==========');
}

main().catch(console.error);

File: generate_explanations.py
#!/usr/bin/env python3
"""
Generate highly specific explanations for all 215 diagnostic test questions.
Uses Claude API to analyze each question and generate detailed explanations.
"""

import json
import os
from anthropic import Anthropic
from supabase import create_client, Client

# Configuration
SUPABASE_URL = "https://rabavobdklnwvwsldbix.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InJhYmF2b2Jka2xud3Z3c2xkYml4Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTk3NjU0NzgsImV4cCI6MjA3NTM0MTQ3OH0.z_1N3TG-cS9Bc1s7UAif91PkIjhBKvicrUqupiNP80Y"

# Initialize clients
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)
anthropic_client = Anthropic(api_key=os.environ.get("ANTHROPIC_API_KEY"))

def load_questions():
    """Load questions from JSON file"""
    with open('/Users/cadenchiang/Desktop/act-prep-react/all_diagnostic_questions.json', 'r') as f:
        return json.load(f)

def get_choice_letter(choices, index):
    """Extract the letter from a choice string"""
    if isinstance(choices[index], str):
        # Try to extract letter from "A. text" format
        if len(choices[index]) > 0 and choices[index][0].isalpha():
            return choices[index][0]
    return chr(65 + index)  # A, B, C, D...

def generate_explanation(question, passage_text=None):
    """Generate a highly specific explanation using Claude API"""
    
    # Build the prompt
    prompt = f"""You are an expert ACT test prep tutor. Generate a HIGHLY SPECIFIC explanation for this question.

SUBJECT: {question['section'].upper()}
QUESTION #{question['question_number']}

"""
    
    if passage_text:
        prompt += f"PASSAGE:\n{passage_text}\n\n"
    
    prompt += f"QUESTION TEXT:\n{question.get('question_text', 'Choose the best answer.')}\n\n"
    
    prompt += "CHOICES:\n"
    for i, choice in enumerate(question['choices']):
        prompt += f"{choice}\n"
    
    prompt += f"\nCORRECT ANSWER: {question['correct_answer']}\n\n"
    
    prompt += """INSTRUCTIONS:
Write a complete explanation with TWO sections in HTML format:

1. FIRST SECTION: Explain why the correct answer is RIGHT
   - Be EXTREMELY specific
   - Reference actual text, numbers, or content from the question
   - Don't say generic things like "provides the most grammatically sound phrasing"
   - DO say specific things like "eliminates the weak 'there are' construction by using 'Of the thousands' to create a subordinate clause"

2. SECOND SECTION: Explain why EACH wrong answer is WRONG
   - Address EVERY incorrect choice
   - Be specific about what makes each one wrong
   - Reference the specific error or issue

FORMAT (use this exact HTML structure):
<div style="line-height: 1.6; margin-bottom: 0.75rem; color: #374151;">
[SPECIFIC explanation of why correct answer works - reference actual content]
</div>

<div>
<strong style="font-size: 0.875rem; color: #6b7280;">Why Other Answers Are Wrong:</strong>
<div style="margin-top: 0.5rem; line-height: 1.6; color: #374151;">
[For each wrong choice, add a div like this:]
<div style="margin-bottom: 0.375rem;"><strong>Choice X:</strong> [Specific reason why it's wrong]</div>
</div>
</div>

Generate the explanation NOW:"""
    
    # Call Claude API
    try:
        message = anthropic_client.messages.create(
            model="claude-sonnet-4-5-20250929",
            max_tokens=2000,
            temperature=0.3,
            messages=[
                {
                    "role": "user",
                    "content": prompt
                }
            ]
        )
        
        explanation = message.content[0].text.strip()
        return explanation
    except Exception as e:
        print(f"ERROR generating explanation for question {question['question_number']}: {e}")
        return None

def update_question_explanation(question, explanation):
    """Update the explanation in the database"""
    section = question['section']
    table_name = f"practice_test_{section}_questions"
    
    try:
        response = supabase.table(table_name).update({
            "explanation": explanation
        }).eq("id", question['id']).execute()
        
        return True
    except Exception as e:
        print(f"ERROR updating question {question['id']}: {e}")
        return False

def main():
    print("Loading questions...")
    questions = load_questions()
    print(f"Loaded {len(questions)} questions\n")
    
    # Group by section
    by_section = {
        'english': [q for q in questions if q['section'] == 'english'],
        'math': [q for q in questions if q['section'] == 'math'],
        'reading': [q for q in questions if q['section'] == 'reading'],
        'science': [q for q in questions if q['section'] == 'science']
    }
    
    print("Breakdown:")
    for section, qs in by_section.items():
        print(f"  {section.capitalize()}: {len(qs)} questions")
    
    print("\n" + "="*80)
    print("Starting explanation generation...")
    print("="*80 + "\n")
    
    total = 0
    successful = 0
    failed = 0
    
    for section, section_questions in by_section.items():
        print(f"\n### Processing {section.upper()} ({len(section_questions)} questions) ###\n")
        
        for i, question in enumerate(section_questions, 1):
            q_num = question['question_number']
            print(f"  [{i}/{len(section_questions)}] Question #{q_num}...", end=" ")
            
            # Generate explanation
            explanation = generate_explanation(question)
            
            if explanation:
                # Update database
                if update_question_explanation(question, explanation):
                    print("OK")
                    successful += 1
                else:
                    print("FAILED TO UPDATE DB")
                    failed += 1
            else:
                print("FAILED TO GENERATE")
                failed += 1
            
            total += 1
            
            # Progress report every 25 questions
            if total % 25 == 0:
                print(f"\n--- PROGRESS: {total}/215 complete ({successful} successful, {failed} failed) ---\n")
    
    print("\n" + "="*80)
    print(f"FINAL RESULTS: {successful}/{total} successful, {failed} failed")
    print("="*80)
    
    # Save sample explanations
    if successful > 0:
        samples = []
        for section_questions in by_section.values():
            if section_questions:
                samples.append(section_questions[0])
            if len(samples) >= 3:
                break
        
        with open('/Users/cadenchiang/Desktop/act-prep-react/sample_explanations.json', 'w') as f:
            json.dump(samples, f, indent=2)
        print("\nSaved sample explanations to sample_explanations.json")

if __name__ == "__main__":
    main()
